### ISSUE TITLE
[Bug] TMap's iterator is different between C# and C++ in UE 4.22

### ISSUE THREAD

**Issue by FJays** [pixeltris/USharp#93]
Hello~
I created a TMap data in C++, and export to C#.
```cpp
UPROPERTY(BlueprintReadWrite)
TMap<int, int> testMap;
```

After I gen the C# code, it created a `TMapReadWrite` in C#.
But the iterator is not fit the UE 4.22 source code(not just delete `KeyOffset` in FScriptMapLayout and FScriptSetLayout).

For example, 
```cpp
testMap.Add(1, 1);
testMap.Add(2, 2);
testMap.Add(3, 3);
testMap.Remove(1);
```

After running the code, the count of the `testMap` is 2. In the VS debug preview window, the first element in `testMap` is Invalid.When I iterate `testMap` in C++ and C#, it's different.
In C++, the iterator can visit all the elements valid(`testMap[2]` and `testMap[3]`). But in C#, the iterator can just visit the invalid element and `testMap[2]`, `testMap[3]` is missing.

I tried to follow the source code, and found that iterator in C++ called `TConstSetBitIterator` in `BitArray.h`, but iterator in C# called `TMapBase.Enumerator`, used `++index` and `map.Count` to iterate, just not fit the source code of UE 4.22.

Looking forward to your reply, thank you!

**Comment by FJays** [comment]
I tried to fix it, maybe it works.
In file `TMap.cs`, I changed some `map.Count` to `MapHelper.GetMaxIndex()`, and used `MapHelper.IsValidIndex(index)` to verify the index :)

**Comment by FJays** [comment]
I made a pull request[https://github.com/pixeltris/USharp/pull/94](pull#94), you can review it.

**Comment by pixeltris** [comment]
Yes it looks like `TSet.cs` will have the same issue. Do you know if the C++ TMap iterator works in constant time (avoids checking invalid indices)? It's hard to follow that C++ code path as it gets passed between many classes.

One interesting thing to note of `FScriptMapHelper` usage in in C++ is that they use `Num()` instead of `GetMaxIndex()` [when looping](https://github.com/EpicGames/UnrealEngine/blob/050fce98785c3925b928842a9b275e5a650e1fea/Engine/Source/Runtime/CoreUObject/Private/UObject/PropertyMap.cpp#L391). I think your usage of `GetMaxIndex()` sounds more correct though.

[`GetAt`](https://github.com/pixeltris/USharp/pull/94/files#diff-15faf73d0236a1c3b1948f283de3412dR120) should check for `IsValidIndex` and return early (or throw an exception) if it fails as the `FScriptMapHelper.GetPairPtr` function doesn't validate the index. `MaxIndex` could also possibly be exposed as a property to help out with iterating via index in user code (if `Count` can be different to max index). Other than that, looks great, thanks.

**Comment by FJays** [comment]
> Yes it looks like `TSet.cs` will have the same issue. If there is a way to do it without checking invalid indices that would be ideal. From what I can tell though, the C++ iterator code is doing somewhat similar index checks:
> 
> * `TMap` iterator creates a `TSet::TBaseIterator`
> * Each iterator loop calls `TSet::TBaseIterator::operator++`, which calls `Id.IsValidId`
> * `TSet::IsValidId` calls `Elements.IsAllocated`
> * `TSparseArray::IsAllocated` returns `AllocationFlags[Index]` (along with a couple of other checks)
> 
> Where as IsValidIndex checks:
> 
> * `AllocationFlags.IsValidIndex(Index) && AllocationFlags[Index];` (where `AllocationFlags` is a `TBitArray`)
> * `TBitArray::IsValidIndex` returns `InIndex >= 0 && InIndex < NumBits`
> 
> Your PR looks great, a couple of possible changes: [`GetAt`](https://github.com/pixeltris/USharp/pull/94/files#diff-15faf73d0236a1c3b1948f283de3412dR120) should check for `IsValidIndex` and return early (or throw an exception) if it fails as the `FScriptMapHelper.GetPairPtr` function doesn't validate the index. `MaxIndex` could also possibly be exposed as a property to help out with iterating via index in user code (and possibly possibly expose an `IsValidIndex()` function?). Other than that, looks great, thanks.

Yeah, you are right.
`GetAt` should throw an `InvalidIndexException`.
But I think exposing an `IsValidIndex()` function is not necessary, because users do not need to know how `TMap` works internally, but use it just like normal `map` with the element's key.

What do you think?
Thank you very much~

**Comment by pixeltris** [comment]
Yea that's very true. Just the `GetAt` change sounds good then.

**Comment by pixeltris** [comment]
Thank. I've merged your PR and made some additional tweaks, along with changes for `TSet`.
